# Launch fungiSMASH on MycoCosm genomes

Use this auxiliary script to launch the fungal version of antiSMASH (previously installed through [conda](https://docs.antismash.secondarymetabolites.org/install/#using-bioconda)) on all genomes downloaded with the MycoCosm genome downloader.

# Requirements

* A working antiSMASH installation (tested on v6)
* A folder structure with fasta + gff files
* A file defining each genome's short name (e.g. JGI's "portal") and paths to the files. Ideally, the `JGI_taxonomy.tsv` file generated by the MycoCosm genome downloader.

# Usage

This script uses the included `antismash_parameters.tsv` file for specific parameters for antiSMASH. Other parameters:

```
usage: launch_fungiSMASH_on_MycoCosm.py [-h] -i INPUTFOLDER -t TAXONOMYFILE -o OUTPUTFOLDER [-p PROCESSES] [-c CPUS]

optional arguments:
  -h, --help            show this help message and exit
  -i INPUTFOLDER, --inputfolder INPUTFOLDER
                        Base input folder
  -t TAXONOMYFILE, --taxonomyfile TAXONOMYFILE
                        Tab-separated file with NCBI accession (column: 'Accession), shortname (column: 'Short name') and relative path to put results (column: Path')
  -o OUTPUTFOLDER, --outputfolder OUTPUTFOLDER
                        Base output directory for antiSMASH results
  -p PROCESSES, --processes PROCESSES
                        Parallel instances of antiSMASH runs. Default: 2
  -c CPUS, --cpus CPUS  'cpus' parameter for each instance of antiSMASH. Default: 2
```
